### PR TITLE
Open iterators reported when cancelling a query during sorting

### PR DESCRIPTION
**Pull request by ehedgehog**
As per JENA-1320. When cancellations in QueryIterSort and QueryIterTopN occur
at the wrong moment, the source iterators (the ones being consumed by Sort
or TopN) are not closed in the usual way but only by the  QueryIteratorCheck
check.

This fix adds additional calls to close() and closeIterator() so that then
usual (non-cancel) paths close the source iterators, and adds extra
handling of the QueryCancelledException so that when it is thrown
through these classes the source iteratots are close(d).

### CODE REVIEW

File: jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterSort.java [hunk 1]
```diff
@@ -89,6 +92,7 @@ public SortedBindingIterator(final QueryIterator qIter) {
             // iterator in a try/finally block, and thus will call
             // close() themselves.
             catch (QueryCancelledException e) {
+		QueryIterSort.this.close();
```
Comment (afs) [inline jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterSort.java]: There are tabs in various plaxces.  Could you follow the convention of using spaces? Display works better. Your tab size != all other tab sizes != Github's handling of tabs.

File: jena-arq/src/test/java/org/apache/jena/sparql/engine/iterator/TestQueryIterSort.java [hunk 2]
```diff
@@ -302,7 +358,10 @@ public void setCallback(Callback callback)
         @Override
         public boolean hasNext() 
         {
-            return delegate.hasNext() ;
+		// self-closing
```
Comment (afs) [inline jena-arq/src/test/java/org/apache/jena/sparql/engine/iterator/TestQueryIterSort.java]: More tabs.

File: jena-arq/src/test/java/org/apache/jena/sparql/engine/iterator/TestQueryIterSort.java [hunk 3]
```diff
@@ -135,6 +133,37 @@ public void testCleanAfterClose()
...
+    @Test public void testCloseClosesSourceIterator() {
+	Context context = new Context() ;
+	ExecutionContext ec = new ExecutionContext(context, (Graph) null, (DatasetGraph) null, (OpExecutorFactory) null);
+	QueryIterSort qis = new QueryIterSort(iterator, comparator, ec);
+	qis.close();
+	assertTrue("source iterator should have been closed", iterator.isClosed());
+    }
+
+        @Test public void testExhaustionClosesSourceIterator() {
+		iterator.setCallback(new Callback()
```
Comment (afs) [inline jena-arq/src/test/java/org/apache/jena/sparql/engine/iterator/TestQueryIterSort.java]: This can be written as `iterator.setCallback(()->{});`

**Comment by afs** [comment]
If you "git rebase" a branch against apache/master, not pull/merge then you won't get the additional merge commits.

**Comment by ehedgehog** [comment]
On 24 April 2017 at 12:54, Andy Seaborne <notifications@github.com> wrote:

> *@afs* approved this pull request.
>

Thanks Andy. What's the next step -- or are we done
for the "open iterators" update?

Chris


-- 
Chris "allusive" Dollin

**Pull request merged**